File: python/proxy_server.py
import socket
import threading

# Function to handle client requests
def handle_client(client_socket):
    # Receive the request from the client
    request = client_socket.recv(1024)

    # Parse the first line of the request to extract the destination URL
    first_line = request.split(b'\n')[0]
    url = first_line.split(b' ')[1]

    # Extract the hostname from the URL
    http_pos = url.find(b'://')  # find position of "://"
    if http_pos == -1:
        temp = url
    else:
        temp = url[(http_pos+3):]  # get the rest of the url

    port_pos = temp.find(b':')  # find the port if present

    # Find the end of the web server
    webserver_pos = temp.find(b'/')
    if webserver_pos == -1:
        webserver_pos = len(temp)

    webserver = ""
    port = -1
    if port_pos == -1 or webserver_pos < port_pos: 
        # Default port
        port = 80
        webserver = temp[:webserver_pos]
    else:  # Specific port
        port = int(temp[(port_pos+1):][:webserver_pos-port_pos-1])
        webserver = temp[:port_pos]

    # Create a socket to connect to the web server
    try:
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.connect((webserver, port))
        server_socket.send(request)  # Send the client's request to the server

        while True:
            # Receive data from the web server
            data = server_socket.recv(4096)

            if len(data) > 0:
                # Send the data to the client
                client_socket.send(data)
            else:
                break
    except socket.error as e:
        print(f"Error: {e}")
    finally:
        # Close the sockets
        client_socket.close()
        server_socket.close()

def start_proxy_server(port):
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.bind(('0.0.0.0', port))
    server.listen(5)
    print(f"[*] Proxy server listening on port {port}")

    while True:
        client_socket, addr = server.accept()
        print(f"[*] Received connection from {addr}")
        client_handler = threading.Thread(target=handle_client, args=(client_socket,))
        client_handler.start()

if __name__ == "__main__":
    port = 8888
    start_proxy_server(port)

File: cpp/proxy_server.cpp
#include <iostream>
#include <string>
#include <thread>
#include <vector>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

void handle_client(int client_socket) {
    char buffer[1024];
    recv(client_socket, buffer, sizeof(buffer), 0);
    
    std::string request(buffer);
    std::string delimiter = " ";
    std::string url = request.substr(request.find(delimiter) + 1);
    url = url.substr(0, url.find(delimiter));

    std::string host;
    int port = 80;
    size_t host_start = url.find("://") + 3;
    size_t port_start = url.find(":", host_start);
    size_t path_start = url.find("/", host_start);

    if (port_start != std::string::npos) {
        host = url.substr(host_start, port_start - host_start);
        port = std::stoi(url.substr(port_start + 1, path_start - port_start - 1));
    } else {
        host = url.substr(host_start, path_start - host_start);
    }

    struct sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    inet_pton(AF_INET, host.c_str(), &server_addr.sin_addr);

    int server_socket = socket(AF_INET, SOCK_STREAM, 0);
    connect(server_socket, (struct sockaddr *)&server_addr, sizeof(server_addr));

    send(server_socket, buffer, sizeof(buffer), 0);
    int bytes_received;
    while ((bytes_received = recv(server_socket, buffer, sizeof(buffer), 0)) > 0) {
        send(client_socket, buffer, bytes_received, 0);
    }

    close(server_socket);
    close(client_socket);
}

int main() {
    int server_socket = socket(AF_INET, SOCK_STREAM, 0);

    struct sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(8888);
    server_addr.sin_addr.s_addr = INADDR_ANY;

    bind(server_socket, (struct sockaddr *)&server_addr, sizeof(server_addr));
    listen(server_socket, 5);

    std::cout << "Proxy server listening on port 8888" << std::endl;

    while (true) {
        int client_socket = accept(server_socket, NULL, NULL);
        std::thread(handle_client, client_socket).detach();
    }

    close(server_socket);
    return 0;
}
